README
MooseX-Registry

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc MooseX::Registry

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=MooseX-Registry

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/MooseX-Registry

    CPAN Ratings
        http://cpanratings.perl.org/d/MooseX-Registry

    Search CPAN
        http://search.cpan.org/dist/MooseX-Registry/


LICENSE AND COPYRIGHT

Copyright (C) 2012 Rob Brown

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for MooseX-Registry

0.02    27 April 2012 - Remove incorrectly included lib

0.01    27 April 2012 - First version




t/lib/Basic.pm
package Basic;

use Moose;

with 'MooseX::Registry::Role';

__PACKAGE__->register('basic-object');

1;


t/lib/MyRole.pm
package MyRole;

use Moose::Role;

with 'MooseX::Registry::Role' => {
	-alias    => { register => 'core_register', lookup => 'core_lookup' },
	-excludes => [ 'register', 'lookup' ]
};

sub register
{
	my $class = shift;
	my $key = shift;
    
	__PACKAGE__->core_register( $key => $class );
}

sub lookup
{
	my $class = shift;
	my $key = shift;
    
	__PACKAGE__->core_lookup( $key );
}

1;



t/lib/MyRegistry.pm
package MyRegistry;

use Moose;

with 'MyRole';



1;


lib/MooseX/Registry/Role.pm
package MooseX::Registry::Role;

use 5.006;

use strict;
use warnings;

use Moose::Role;

=head1 NAME

MooseX::Registry::Role - The role for MooseX::Registry.

=head1 VERSION

Version 0.02

=cut

our $VERSION = '0.02';

$VERSION = eval $VERSION;

=head1 SYNOPSIS

 package Long::Name::I::Want::To::Change::Later;
 
 with MooseX::Registry::Role;
 
 __PACKAGE__->register( 'nice-name-v2');
 
 1;

 use MooseX::Registry;
  
 $class = MooseX::Registry->lookup( 'nice-name-v2' );
 
 my $instance = $class->new;

=head1 METHODS

=head2 register

Adds an entry in the registry for this class name with a label.

=cut

sub register
{
	if ( scalar @_ == 2 )
	{
		# no namespace
		
		my $class     = shift;
		my $key       = shift;

		MooseX::Registry->instance->_hash->{ $key } = $class;
	}
	else
	{
		my $namespace = shift;
		my $key       = shift;
		my $class     = shift;

		my $hash = MooseX::Registry->instance->_hash;
	
		$hash->{ $namespace } = {} unless exists $hash->{ $namespace };
	
		$hash->{ $namespace }->{ $key } = $class;
	}
}

=head2 lookup

Looks up a label to find the original class name.

=cut

sub lookup
{
	my $namespace = shift;
	my $key       = shift;

	my $hash = MooseX::Registry->instance->_hash;

	return $hash->{ $namespace }->{ $key } if exists $hash->{ $namespace } && ref $hash->{ $namespace } eq 'HASH' && exists $hash->{ $namespace }->{ $key };
	
	return $hash->{ $key } if exists $hash->{ $key };
	
	return undef;
}

=head1 AUTHOR

Rob Brown, C<< <rob at intelcompute.com> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2012 Rob Brown.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;



lib/MooseX/Registry/Register.pm
package MooseX::Registry;

use 5.006;

use strict;
use warnings;

use Moose::Role;

=head1 NAME

MooseX::Registry - Stay away from namespaces and class names with a Moose registry.

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';

$VERSION = eval $VERSION;

=head1 SYNOPSIS

Put your classes in the registry...

 package MyApp::Class::Name::v2_3;
 
 with MooseX::Registry::Role;
 
 __PACKAGE__->register( 'myclass-v2.3' );

Look them up later...

 with MooseX::Role::Registry;
  
 $class = $self->lookup( 'myclass-v2.3' );
 
 my $instance = $class->new;




=head1 METHODS

=head2 register

=cut

sub function1 {
}

=head2 function2

=cut

sub function2 {
}

=head1 AUTHOR

Rob Brown, C<< <rob at intelcompute.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-moosex-role-registry at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=MooseX-Role-Registry>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc MooseX::Role::Registry


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=MooseX-Role-Registry>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/MooseX-Role-Registry>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/MooseX-Role-Registry>

=item * Search CPAN

L<http://search.cpan.org/dist/MooseX-Role-Registry/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2012 Rob Brown.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1; # End of MooseX::Role::Registry


lib/MooseX/Registry.pm
package MooseX::Registry;

use 5.006;

use strict;
use warnings;

use MooseX::Singleton;

with 'MooseX::Registry::Role';

=head1 NAME

MooseX::Registry - Hide namespaces and class names with a Moose registry.

=head1 VERSION

Version 0.02

=cut

our $VERSION = '0.02';

$VERSION = eval $VERSION;

has _hash => ( is      => 'rw',
               isa     => 'HashRef',
               default => sub { {} },
             );

=head1 SYNOPSIS

Put your classes in the registry...

 package Long::Name::I::Want::To::Change::Later;
 
 with MooseX::Registry::Role;
 
 __PACKAGE__->register( 'nice-name-v2');
 
 1;

Look them up later...

 use MooseX::Registry;
  
 $class = MooseX::Registry->lookup( 'nice-name-v2' );
 
 my $instance = $class->new;

=head1 DESCRIPTION

If you have to deal with collisions, you can extend the role to introduce namespacing separation by usinmg your own role...

 package MyRole;

 use Moose::Role;
 
 with 'MooseX::Registry::Role' => {
         -alias    => { register => 'core_register', lookup => 'core_lookup' },
         -excludes => [ 'register', 'lookup' ]
 };
 
 sub register
 {
         my $class = shift;
         my $key = shift;
 
         __PACKAGE__->core_register( $key => $class );
 }
 
 sub lookup
 {
         my $class = shift;
         my $key = shift;
 
         __PACKAGE__->core_lookup( $key );
 }

Now consume your own role.  The important bit is the call to C<core_register>.

You can C<alias> the original methods however you choose, and also name your own C<register> and C<lookup> methods inline with your own project.
         
=head1 AUTHOR

Rob Brown, C<< <rob at intelcompute.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-moosex-role-registry at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=MooseX-Registry>.  I will be notified, and then you will
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc MooseX::Registry


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=MooseX-Registry>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/MooseX-Registry>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/MooseX-Registry>

=item * Search CPAN

L<http://search.cpan.org/dist/MooseX-Registry/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2012 Rob Brown.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1;

